api/src/main/java/org/apache/iceberg/catalog/SupportsCatalogTransactions.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Catalog methods for working with catalog-level transactions.
+ *
+ * <p>Catalog implementations are not required to support catalog-level transactional state.
+ * If they do, they may support one or more {@code IsolationLevel}s and one or more
+ * {@code LockingMode}s.
+ */
+public interface SupportsCatalogTransactions {
+
+  /**
+   * The level of isolation for a transaction.
+   */
+  enum IsolationLevel {
+
+    /**
+     * Only read committed data. Reading the same table multiple times
+     * may result in different committed reads for each read.
+     */
+    READ_COMMITTED,
+
+    /**
+     * Only read committed data. Reading the same table multiple times will
+     * result in the same view of those tables.
```

Comment:
I'm thinking of `REPEATED_READ` as we have been discussing "snapshot" isolation: your changes are made using a consistent snapshot of a table, and as long as those changes can still be committed as though that snapshot were still current, the commit will succeed. That fits with the idea that multiple reads get the same version of the table. That seems to fit with the Wikipedia description for write behavior: "Write skew is possible at this isolation level" because two writers may be based on the same version and both commit successfully, not detecting that a region of data has been changed by the other write.

Reply:
I was talking with Jacques on Slack and I think the difference between this "repeatable read" and the "snapshot" that we use for single-table transactions is that with multiple tables, snapshot implies that there is a point in time when all tables are in the starting state. Catalogs may not guarantee that there is happened-before relationship across tables, so repeatable read is used instead of snapshot.

However if I understand correctly, the same write guarantees are used for both "repeatable read" and "snapshot": there is no check that newer commits have not occurred that would change the result of the operation.

Reply:
Thanks for the explanations! The new definition looks very clear to me. I can roughly imagine how it plays out in Glue and Hive with this interface, and I can put up an initial draft after this is merged.